Ntran/ingest data for querier

NGA-TRAN opened the pull request:
Closes #

Describe your proposed changes here.

- [ ] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

carols10cents commented on ingester/src/data.rs:
```diff
@@ -868,6 +931,25 @@ impl DataBuffer {
+    pub fn buffer_and_snapshots(&self) -> Result<Vec<Arc<SnapshotBatch>>> {
+        let mut snapshots = vec![];
+        // copy the buffer to a snapshot
+        let buffer_sanpshot = self
+            .copy_buffer_to_snapshot()
+            .context(BufferToSnapshotSnafu)?;
+        if let Some(snapshot) = buffer_sanpshot {
+            snapshots.push(snapshot);
+        }
```
Typo in the variable name, might make ctrl-f for this unexpectedly fail:

```suggestion
        let buffer_snapshot = self
            .copy_buffer_to_snapshot()
            .context(BufferToSnapshotSnafu)?;
        if let Some(snapshot) = buffer_snapshot {
            snapshots.push(snapshot);
        }
```

carols10cents commented on ingester/src/data.rs:
```diff
@@ -868,6 +931,25 @@ impl DataBuffer {
+        }
+
+        // other existing snapshots
+        let mut other_snapshots: Vec<_> = self.snapshots.iter().map(Arc::clone).collect();
+        snapshots.append(&mut other_snapshots);
```
This can use `extend` and avoid the allocation of the `Vec`:

```suggestion
        snapshots.extend(self.snapshots.iter().cloned());
```

carols10cents commented on ingester/src/data.rs:
```diff
@@ -909,6 +991,25 @@ impl DataBuffer {
+        let mut queryable_batch = self.persisting.as_ref().unwrap().data.copy(sequence_number);
+
+        // Add new tombstones if any
+        queryable_batch.add_tombestones(&mut self.deletes_during_persisting.clone());
```
Typo in this method name:

```suggestion
        queryable_batch.add_tombstones(&mut self.deletes_during_persisting.clone());
```

carols10cents commented on ingester/src/querier_handler.rs:
```diff
@@ -1,74 +1,231 @@
     ))]
     ExecutePlan { source: DataFusionError },
+
+    #[snafu(display("Error collecting a stream of record bacthes: {}", source))]
```
Typo:

```suggestion
    #[snafu(display("Error collecting a stream of record batches: {}", source))]
```

carols10cents commented on ingester/src/querier_handler.rs:
```diff
@@ -1,74 +1,231 @@
+    CollectStream { source: DataFusionError },
+
+    #[snafu(display(
+        "No Table Data found for the given sequencer id {}, namspace name {}, table name {}", sequencer_id.get(), namespace_name, table_name
```
Typo:

```suggestion
        "No Table Data found for the given sequencer id {}, namespace name {}, table name {}", sequencer_id.get(), namespace_name, table_name
```

carols10cents commented on ingester/src/querier_handler.rs:
```diff
@@ -1,74 +1,231 @@
+    #[snafu(display("Error snapshotting non-persisting data: {}", source))]
+    SnapshotNonPersistingData { source: data::Error },
+
+    #[snafu(display("Error concating same-schema record bacthes: {}", source))]
```
Typo:

```suggestion
    #[snafu(display("Error concating same-schema record batches: {}", source))]
```

carols10cents commented on ingester/src/querier_handler.rs:
```diff
@@ -1,74 +1,231 @@
+        request.namespace.as_str(),
+        request.table.as_str(),
+    );
+    if table_data.is_none() {
+        return Err(Error::TableNotFound {
+            sequencer_id: request.sequencer_id,
+            namespace_name: request.namespace.clone(),
+            table_name: request.table.clone(),
+        });
+    }
+
+    // ------------------------------------------------
+    // Accumulate data from each partition
+    let table_data = table_data.unwrap();
```
This `unwrap`, even though it's ok because of the `return` if `table_data` is none above, takes a minute to read and verify that this is ok and won't actually panic. I think this would be clearer if we add `use snafu::OptionExt;` at the top and then do:

```suggestion
    let table_data = table_data.context(TableNotFoundSnafu {
        sequencer_id: request.sequencer_id,
        namespace_name: &request.namespace,
        table_name: &request.table,
    })?;

    // ------------------------------------------------
    // Accumulate data from each partition
```

This also lets snafu take care of cloning when needed.

carols10cents commented on ingester/src/querier_handler.rs:
```diff
@@ -1,74 +1,231 @@
+    let partition_keys = table_data.partition_keys();
+
+    // Make Filters
+    let mut selection: Selection<'_> = Selection::All;
+    let mut sel = vec![];
+    if !request.columns.is_empty() {
+        for col in &request.columns {
+            sel.push(col.as_str());
+        }
+        selection = Selection::Some(&sel);
+    }
```
This is potentially doing multiple allocations of Vec as it grows; better would be using a `collect` to allocate the exact size once. This is a little tricky with `Selection::Some` needing a slice, but I think this is a little better at least:

```suggestion
    let selection_columns: Vec<_> = request.columns.iter().map(String::as_str).collect();
    let selection = if selection_columns.is_empty() {
        Selection::All
    } else {
        Selection::Some(&selection_columns)
    };
```

NGA-TRAN replied:
Yeah, I got compile error when trying to create a slice :)

carols10cents commented on ingester/src/querier_handler.rs:
```diff
@@ -1,74 +1,231 @@
+        }
+        selection = Selection::Some(&sel);
+    }
+    let mut predicate = Predicate::default();
+    if let Some(pred) = &request.predicate {
+        predicate = (*pred).clone();
+    }
```
This can be:

```suggestion
    let predicate = request.predicate.clone().unwrap_or_default();
```

carols10cents commented on ingester/src/querier_handler.rs:
```diff
@@ -1,74 +1,231 @@
+    // Make a stream for this batch
+    let dummy_metrics = ExecutionPlanMetricsSet::new();
+    let mem_metrics = MemTrackingMetrics::new(&dummy_metrics, 0);
+    let mut stream_batch = vec![];
+    if let Some(batch) = batch {
+        stream_batch.push(Arc::new(batch));
+    }
```
This can be:

```suggestion
    let stream_batch = batch.map(|b| vec![Arc::new(b)]).unwrap_or_default();
```

carols10cents commented on ingester/src/query.rs:
```diff
@@ -89,6 +89,32 @@ impl QueryableBatch {
+    }
+
+    /// Add more tombstones
+    pub fn add_tombestones(&mut self, deletes: &mut Vec<Tombstone>) {
```
Typo in method name:

```suggestion
    pub fn add_tombstones(&mut self, deletes: &mut Vec<Tombstone>) {
```

carols10cents reviewed (changes requested):
Some typos and places that could be a bit rustier :)

NGA-TRAN reviewed (commented):
Thanks @carols10cents for the suggestions. Very helpful

NGA-TRAN commented:
@carols10cents : I have added tests for many scenarios and addressed your coding comments. This is ready for review. 
I am adding more scenario tests for deletes and persisting but the code will stay the same unless those tests discover bugs

NGA-TRAN commented:
@carols10cents @pauldix @crepererum : I am happy with all the tests. This is ready for final review

pauldix commented on ingester/src/data.rs:
```diff
@@ -836,6 +911,19 @@ impl DataBuffer {
+    }
+
+    /// Returns snapshot of the buffer but keep data in the buffer
+    pub fn copy_buffer_to_snapshot(
```
What's the reason for doing this vs. just snapshotting it and moving it over there?

NGA-TRAN replied:
@pauldix As we discussed earlier, by doing this, we only take read lock, not write lock.

pauldix reviewed (commented):
A few questions on this

NGA-TRAN commented on ingester/src/data.rs:
```diff
@@ -894,6 +980,24 @@ impl DataBuffer {
+            .copy_buffer_to_snapshot()
+            .context(BufferToSnapshotSnafu)?;
+        if let Some(snapshot) = buffer_snapshot {
+            snapshots.push(snapshot);
```
Good catch.

shepmaster commented on ingester/src/test_util.rs:
```diff
@@ -542,3 +554,336 @@ pub async fn create_batches_with_influxtype_same_columns_different_type() -> Vec
+    let partitions = make_partitions(two_partitions, loc, seq_id, data_table_id, TEST_TABLE);
+
+    // Two tables: one empty and one with data of one or two partitions
+    let mut tables: BTreeMap<String, Arc<TableData>> = BTreeMap::default();
```
No need for the type annotation (here and other places)

```suggestion
    let mut tables = BTreeMap::default();
```

I usually use `BTreeMap::new()` when I care to specifically make it a `BTreeMap` or I'll use `Default::default()` when I don't care and just want a fresh value.

NGA-TRAN replied:
I will use `BTreeMap::new()` then

pauldix commented on ingester/src/data.rs:
```diff
@@ -894,6 +980,24 @@ impl DataBuffer {
+        let buffer_snapshot = self
+            .copy_buffer_to_snapshot()
+            .context(BufferToSnapshotSnafu)?;
+        if let Some(snapshot) = buffer_snapshot {
+            snapshots.push(snapshot);
+        }
```
TIL that's a cool one 😄

NGA-TRAN commented on ingester/src/query.rs:
```diff
@@ -67,28 +67,49 @@ pub type Result<T, E = Error> = std::result::Result<T, E>;
+    /// We need this new QueryableBatch because we want to add more tombstones as needed
+    /// but do not want to chnage the tombstone list of  original QueryableBatch
+    pub fn copy(&self) -> Self {
+        Self {
+            data: self.data.clone(),
+            delete_predicates: self.delete_predicates.clone(),
+            table_name: self.table_name.clone(),
+        }
```
Good catch. The code at first was
```rust
Self {
             data:  SOMETHING,
             delete_predicates: self.delete_predicates.clone(),
             table_name: self.table_name.clone(),
         }
```
When I replaced SOMETHING with `self.data.clone()`, I did not notice it was simply a clone :)

NGA-TRAN reviewed (commented):
Thanks @shepmaster. Super helpful for improving my Rust skills

NGA-TRAN commented:
@carols10cents @pauldix @shepmaster I have addressed all comments

shepmaster commented on ingester/src/test_util.rs:
```diff
@@ -542,3 +554,336 @@ pub async fn create_batches_with_influxtype_same_columns_different_type() -> Vec
+#[allow(missing_docs)]
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum DataLocation {
+    Buffer = 1,
```
If the values do not matter, do not put them:

```rust
pub enum DataLocation {
    Buffer,
    Snapshot,
}
```

NGA-TRAN replied:
Done

shepmaster replied:
> a bitflags type...

For completeness, here's what that might look like...

```rust
use bitflags; // 1.3.2

bitflags::bitflags! {
    struct DataLocation: u8 {
        const BUFFER = 0b0001;
        const SNAPSHOT = 0b0010;
        const PERSISTING = 0b0100;
    }
}
```

Then remove

```rust
match loc {
    DataLocation::Persisting
    | DataLocation::SnapshotPersisting
    | DataLocation::BufferSnapshotPersisting
    | DataLocation::BufferPersisting => {
        // Move group 1 data to persisting
        p1.snapshot_to_persisting_batch(sequencer_id, table_id, partition_id, table_name);
    }
    DataLocation::BufferSnapshot | DataLocation::Snapshot => {
        // move group 1 data to snapshot
        p1.snapshot().unwrap();
    }
    _ => {} // keep it in buffer
}
```

and replace with

```rust
if loc.contains(DataLocation::PERSISTING) {
    // Move group 1 data to persisting
    p1.snapshot_to_persisting_batch(sequencer_id, table_id, partition_id, table_name);
} else if loc.contains(DataLocation::SNAPSHOT) {
    // move group 1 data to snapshot
    p1.snapshot().unwrap();
} else {
    // keep it in buffer
}
```

NGA-TRAN replied:
I am trying to replace with this but I do need all 7 types of the enums

shepmaster replied:
I don't know that you _need_ to replace it: if the enums work that's fine. This was mostly me musing and then realizing I hadn't explained that musing in enough detail.

If you wanted to move forward with it, you could construct combinations via `|`: `DataLocation::PERSISTING | DataLocation::SNAPSHOT`.

NGA-TRAN replied:
I noticed since they are just used for tests, it did not matter but since I want to learn about bitflags, I replaced anyway :)

NGA-TRAN commented:
I think I am happy with this now. 
@carols10cents I need you to unlock this to merge

shepmaster commented on ingester/src/test_util.rs:
```diff
@@ -590,27 +586,38 @@ pub const TEST_PARTITION_1: &str = "test+partition_1";
+    ///       . snapshot + persisting
+    ///       . buffer + snapshot + persisting
+    ///    . If the second partittion exists, it only has data in its buffer
+    pub struct DataLocation: u16 {
+        ///
+        const BUFFER = 0b00000001;
+        ///
+        const SNAPSHOT = 0b00000010;
+        ///
+        const PERSISTING = 0b00000100;
+        ///
+        const BUFFER_SNAPSHOT = 0b0001000;
+        ///
+        const BUFFER_PERSISTING = 0b0010000;
+        ///
+        const SNAPSHOT_PERSISTING = 0b0100000;
+        ///
+        const BUFFER_SNAPSHOT_PERSISTING = 0b1000000;
+    }
```
The point of a bitflags type is that it models a set of booleans. This should be

```suggestion
    pub struct DataLocation: u8 {
        const BUFFER = 0b001;
        const SNAPSHOT = 0b010;
        const PERSISTING = 0b100;
        
        // If you wanted the precomposed aliases:
        const BUFFER_SNAPSHOT = Self::BUFFER.bits | Self:: SNAPSHOT.bits;
        const BUFFER_PERSISTING = Self::BUFFER.bits | Self:: PERSISTING.bits;
        const SNAPSHOT_PERSISTING = Self::SNAPSHOT.bits | Self:: PERSISTING.bits;
        const BUFFER_SNAPSHOT_PERSISTING =  Self::BUFFER.bits | Self::SNAPSHOT.bits | Self:: PERSISTING.bits;
    }
```

NGA-TRAN replied:
Ah, I see. Silly me. I did ask myself why we did not use 0b011. Changing it

shepmaster commented on ingester/src/test_util.rs:
```diff
@@ -656,6 +664,8 @@ pub fn make_ingester_data(two_partitions: bool, loc: DataLocation) -> crate::dat
 
     // Ingester data that inlcudes one sequencer/shard
 
+    use std::collections::BTreeMap;
```
This `use` seems oddly placed; why not put it at the top of the file? Then you can remove the fully-qualified usage above.

NGA-TRAN replied:
I created this file to include functions for testing only but do not want to put them inside another struct. Because of that I need to add `#[cfg(test)]` on top of each function and as a consequence, it does not accept `use` on top on the file if they are parameter or return types of the function

alamb replied:
putting the `use` at the beginning of the function might be less jarring for future readers, perhaps:

```rust
#[cfg(test)]
pub fn make_ingester_data(two_partitions: bool, loc: DataLocation) -> crate::data::IngesterData {
    use std::collections::BTreeMap;
    ...
```

NGA-TRAN replied:
Done

shepmaster commented on ingester/src/test_util.rs:
```diff
@@ -884,21 +910,19 @@ fn make_first_partition_data(
-        _ => {} // keep it in buffer
-    }
+
+    if loc.contains(DataLocation::PERSISTING)
+        || loc.contains(DataLocation::SNAPSHOT_PERSISTING)
+        || loc.contains(DataLocation::BUFFER_PERSISTING)
+        || loc.contains(DataLocation::BUFFER_SNAPSHOT_PERSISTING)
+    {
```
This would just be 
```suggestion
    if loc.contains(DataLocation::PERSISTING) {
```

shepmaster commented on ingester/src/test_util.rs:
```diff
@@ -884,21 +910,19 @@ fn make_first_partition_data(
+    {
+        // Move group 1 data to persisting
+        p1.snapshot_to_persisting_batch(sequencer_id, table_id, partition_id, table_name);
+    } else if loc.contains(DataLocation::BUFFER_SNAPSHOT) || loc.contains(DataLocation::SNAPSHOT) {
```
```suggestion
    } else if loc.contains(DataLocation::SNAPSHOT) {
```

shepmaster commented on ingester/src/test_util.rs:
```diff
@@ -909,21 +933,22 @@ fn make_first_partition_data(
-        _ => {} // keep it in buffer
-    }
+
+    if loc.contains(DataLocation::PERSISTING)
+        || loc.contains(DataLocation::SNAPSHOT_PERSISTING)
+        || loc.contains(DataLocation::BUFFER_PERSISTING)
+        || loc.contains(DataLocation::BUFFER_SNAPSHOT_PERSISTING)
+    {
```
```suggestion
    if loc.contains(DataLocation::PERSISTING) {
```

NGA-TRAN commented:
@shepmaster I think I got bitflags this time :)

alamb commented on ingester/Cargo.toml:
```diff
@@ -47,5 +48,6 @@ tokio-util = { version = "0.7.0" }
 
 [dev-dependencies]
-mutable_batch_lp = { path = "../mutable_batch_lp" }
+bitflags = {version = "1.3.2"}
```
for anyone else following along, `bitflags` is not a new dependency (already exists in our stack somewhere)

alamb commented on ingester/src/querier_handler.rs:
```diff
@@ -1,74 +1,219 @@
+    }
+
+    // ------------------------------------------------
+    // Combine record batches into one batch and padding null values as needed
```
```suggestion
    // Combine record batches into one batch and pad null values as needed
```

alamb commented on ingester/src/querier_handler.rs:
```diff
@@ -1,74 +1,219 @@
+    // ------------------------------------------------
+    // Combine record batches into one batch and padding null values as needed
+
+    // Schema of all record batches after mergeing
```
```suggestion
    // Schema of all record batches after merging
```

alamb commented on ingester/src/querier_handler.rs:
```diff
@@ -1,74 +1,219 @@
+        let queryable_batch =
+            QueryableBatch::new(&request.table, filter_not_applied_batches, vec![]);
+
+        let record_batch = run_query(
```
as a future optimization it might be cool to actually "stream" the results together here (as written this will run the queryable batches one after the other sequentially and buffer the results in memory). We could probably combine the streams with a `CoalescePartitionsExec` https://github.com/apache/arrow-datafusion/blob/master/datafusion/src/physical_plan/coalesce_partitions.rs#L48

alamb commented on ingester/src/data.rs:
```diff
@@ -540,6 +578,32 @@ impl TableData {
+        }
+    }
+
+    ///
```
Missing comment?

NGA-TRAN replied:
Done

alamb commented on ingester/src/querier_handler.rs:
```diff
@@ -221,4 +367,202 @@ mod tests {
+            DataLocation::SNAPSHOT_PERSISTING,
+            DataLocation::PERSISTING,
+        ] {
+            let scenario = Arc::new(make_ingester_data_with_tombstones(*loc).await);
```
👍

alamb commented on ingester/src/test_util.rs:
```diff
@@ -572,3 +572,381 @@ pub async fn create_batches_with_influxtype_same_columns_different_type() -> Vec
+    table_id: TableId,
+    table_name: &str,
+) -> std::collections::BTreeMap<String, Arc<crate::data::PartitionData>> {
+    // In-memory data inlcudes these rows but split between 4 groups go into
```
```suggestion
    // In-memory data includes these rows but split between 4 groups go into
```

alamb commented on ingester/src/test_util.rs:
```diff
@@ -572,3 +572,381 @@ pub async fn create_batches_with_influxtype_same_columns_different_type() -> Vec
+    table_id: TableId,
+    table_name: &str,
+) -> std::collections::BTreeMap<String, Arc<crate::data::PartitionData>> {
+    // In-memory data inlcudes these rows but split between 4 groups go into
```
```suggestion
    // In-memory data includes these rows but split between 4 groups go into
```

alamb reviewed (approved):
I reviewed the running of queries in `query_handler` and tests -- very impressive. Nice work @NGA-TRAN

carols10cents commented on Cargo.lock:
```diff
@@ -69,9 +69,9 @@ dependencies = [
 [[package]]
 name = "anyhow"
-version = "1.0.55"
+version = "1.0.53"
```
A bunch of these changes look like downgrades from Marco's PR that just upgraded everything-- could you try rebasing on main and whenever there's a conflict for `Cargo.lock`, run:

```
git checkout --ours Cargo.lock
cargo check
git add Cargo.lock
git rebase --continue
```

and then force push the branch to see if that puts fewer downgrades in the Cargo.lock? I'm also happy to take care of this if you'd like!

NGA-TRAN replied:
@carols10cents : the Cargo.lock is fixed

carols10cents commented on ingester/src/query.rs:
```diff
@@ -67,28 +67,40 @@ pub type Result<T, E = Error> = std::result::Result<T, E>;
     }
 
+    /// Add more tombstones
+    pub fn add_tombstones(&mut self, deletes: &mut Vec<Tombstone>) {
```
It doesn't look like this function needs a mutable reference to a vec anymore, so I think this can be:

```suggestion
    pub fn add_tombstones(&mut self, deletes: &[Tombstone]) {
```

and then the call site can eliminate a `clone`, commenting there too...

carols10cents commented on ingester/src/data.rs:
```diff
@@ -956,6 +1069,22 @@ impl DataBuffer {
+        let mut queryable_batch = (*persisting.data).clone();
+
+        // Add new tombstones if any
+        queryable_batch.add_tombstones(&mut self.deletes_during_persisting.clone());
```
To go with the change in ingester/src/query.rs, this can be:

```suggestion
        queryable_batch.add_tombstones(&self.deletes_during_persisting);
```

carols10cents commented on ingester/src/data.rs:
```diff
@@ -956,6 +1069,22 @@ impl DataBuffer {
     }
 
+    /// Return a QueryableBatch of the persisting batch after applying new tombstones
+    pub fn get_persisting_data(&self) -> Result<Option<QueryableBatch>> {
```
I don't see anywhere that this function would return an `Err` type; does it still need to return a `Result` or could this be:

```suggestion
    pub fn get_persisting_data(&self) -> Option<QueryableBatch> {
```

?

NGA-TRAN replied:
Yes, the result is using to return the right error message to the cline (which is Querier). This function invokes the one you commented above https://github.com/influxdata/influxdb_iox/pull/3825#discussion_r816044742

carols10cents commented on ingester/src/test_util.rs:
```diff
@@ -572,3 +572,386 @@ pub async fn create_batches_with_influxtype_same_columns_different_type() -> Vec
     batches
 }
+
+///
```
If you don't think anything in this module needs documentation comments, you can turn off the lint for the whole module by adding:

```
#![allow(missing_docs)]
```

to the top of this file. Either that or adding explanations for these values would be better than empty comments ;)

carols10cents reviewed (changes requested):
This is getting super close! Just a few more things!

NGA-TRAN commented on ingester/src/query.rs:
```diff
@@ -67,28 +67,40 @@ pub type Result<T, E = Error> = std::result::Result<T, E>;
+                    &tombstone.max_time.get().to_string(),
+                    &tombstone.serialized_predicate,
+                )
+                .expect("Error building delete predicate"),
```
Good catch. I will see if I can return a Result for this function

NGA-TRAN replied:
@carols10cents : Now I remember why I did this on purpose. We only support very few types of tombstones. Because of that, if something is wrong here, I want to panic for us to look into it immediately during massive tests to avoid this kind of error in product.

NGA-TRAN commented on ingester/src/test_util.rs:
```diff
@@ -572,3 +572,386 @@ pub async fn create_batches_with_influxtype_same_columns_different_type() -> Vec
+
+    // Two tables: one empty and one with data of one or two partitions
+    let mut tables: std::collections::BTreeMap<String, Arc<crate::data::TableData>> =
+        std::collections::BTreeMap::new();
```
Done

NGA-TRAN commented:
@carols10cents : I have addressed all but one which is putting `use ...` on top of the test_util.rs. The functions are created for test only but not in a module. Because of that, the `use` has to be inside each function which will also be a lot. I decided to use path directly instead of many uses.
It is ready for review again :)

carols10cents commented:
> The functions are created for test only but not in a module.

Ah, I didn't realize that was the issue. If we make the whole `test_util` module test-only, then we can bring these items into the `test_util` module's scope-- I've added a commit to your PR doing that because I tested it out locally to be sure it would work. Please take a look at the last commit and if you like it, I think this is ready to merge!

NGA-TRAN commented:
Thanks so much @carols10cents for making the test_util a module which was a lot simpler than I had thought  😄
